fastcrypto/src/tests/bls12381_tests.rs

```diff
@@ -541,3 +543,134 @@ fn test_signature_aggregation() {
     let blst_signatures: Vec<_> = blst_keypairs.iter().map(|key| key.sign(msg)).collect();
     assert!(BLS12381AggregateSignature::aggregate(&blst_signatures).is_err());
 }
+
+// Arbitrary implementations for the proptests
+fn arb_keypair() -> impl Strategy<Value = BLS12381KeyPair> {
+    any::<[u8; 32]>()
+        .prop_map(|seed| {
+            let mut rng = StdRng::from_seed(seed);
+            BLS12381KeyPair::generate(&mut rng)
+        })
+        .no_shrink()
+}
+
+prop_compose! {
+    fn valid_signature(pk: BLS12381PrivateKey)
+                      (msg in any::<[u8; 32]>()) -> ([u8; 32], BLS12381Signature) {
+            (msg, pk.sign(&msg))
+    }
+}
+
+prop_compose! {
+  fn maybe_valid_sig(pk: BLS12381PrivateKey)
+                    (disc in bool::arbitrary(),
+                    (msg, sig) in valid_signature(pk))
+                       -> ([u8; 32], BLS12381Signature) {
+    if disc {
+      (msg, sig)
+    } else {
+      let mut rng = StdRng::from_seed([0; 32]);
+      let mut msg = msg;
+      rng.fill_bytes(&mut msg);
+      (msg, sig)
+    }
+  }
+}
+
+fn arb_sig_triplet() -> impl Strategy<Value = (BLS12381PublicKey, [u8; 32], BLS12381Signature)> {
+    arb_keypair()
+        .prop_flat_map(|kp| {
+            let pk: BLS12381PublicKey = kp.public().clone();
+            (Just(pk), maybe_valid_sig(kp.private()))
+        })
+        .prop_flat_map(|(pk, (msg, sig))| (Just(pk), Just(msg), Just(sig)))
+        .no_shrink()
+}
+
+const BLS_MAX_SIGNATURES: usize = 100;
+
+fn aggregate_treewise(sigs: &[BLS12381Signature]) -> BLS12381AggregateSignature {
```

Comment:
Interesting, when is this useful, to ensure that different aggregation strategies are aligned?

Reply:
Oh, I just wanted to not limit my testing to `add_signature`. Eventually this will be useful in tests exercising https://eprint.iacr.org/2021/377.pdf section 3.2.1.